### PR TITLE
Use ::class statement instead of string

### PR DESCRIPTION
**Pull request by MorrisJobke**
Used a PHPStorm inspection to do that. (it also proposed this ;))

### CODE REVIEW

**Comment by rullzer** [comment]
CI is not happy

File: apps/dav/lib/CalDAV/Search/SearchPlugin.php [hunk 1]
```diff
@@ -78,7 +78,7 @@ public function initialize(Server $server) {
 		$server->on('report', [$this, 'report']);
 
 		$server->xml->elementMap['{' . self::NS_Nextcloud . '}calendar-search'] =
-			'OCA\\DAV\\CalDAV\\Search\\Xml\\Request\\CalendarSearchReport';
+			Xml\Request\CalendarSearchReport::class;
```
Comment (nickvergessen) [inline apps/dav/lib/CalDAV/Search/SearchPlugin.php]: I hate these partial imports

File: apps/dav/lib/DAV/Sharing/Plugin.php [hunk 2]
```diff
@@ -100,8 +100,8 @@ function getPluginName() {
 	 */
 	function initialize(Server $server) {
 		$this->server = $server;
-		$this->server->xml->elementMap['{' . Plugin::NS_OWNCLOUD . '}share'] = 'OCA\\DAV\\DAV\\Sharing\\Xml\\ShareRequest';
-		$this->server->xml->elementMap['{' . Plugin::NS_OWNCLOUD . '}invite'] = 'OCA\\DAV\\DAV\\Sharing\\Xml\\Invite';
+		$this->server->xml->elementMap['{' . Plugin::NS_OWNCLOUD . '}share'] = Xml\ShareRequest::class;
```
Comment (nickvergessen) [inline apps/dav/lib/DAV/Sharing/Plugin.php]: same here

**Comment by MorrisJobke** [comment]
I fixed the merge conflict, the partial imports as well as the CI issue.

File: lib/private/Server.php [hunk 3]
```diff
@@ -242,7 +242,7 @@ public function __construct($webRoot, \OC\Config $config) {
 
 		$this->registerService('SystemTagManagerFactory', function (Server $c) {
 			$config = $c->getConfig();
-			$factoryClass = $config->getSystemValue('systemtags.managerFactory', '\OC\SystemTag\ManagerFactory');
+			$factoryClass = $config->getSystemValue('systemtags.managerFactory', SystemTag\ManagerFactory::class);
```
Comment (ChristophWurst) [inline lib/private/Server.php]: I guess @nickvergessen won't like this one neither 😉

File: lib/private/Server.php [hunk 4]
```diff
@@ -317,27 +317,27 @@ public function __construct($webRoot, \OC\Config $config) {
 		$this->registerService(Store::class, function (Server $c) {
 			$session = $c->getSession();
 			if (\OC::$server->getSystemConfig()->getValue('installed', false)) {
-				$tokenProvider = $c->query('OC\Authentication\Token\IProvider');
+				$tokenProvider = $c->query(Authentication\Token\IProvider::class);
```
Comment (ChristophWurst) [inline lib/private/Server.php]: same

File: lib/private/Server.php [hunk 5]
```diff
@@ -1020,7 +1020,7 @@ public function __construct($webRoot, \OC\Config $config) {
 
 		$this->registerService(\OCP\Share\IManager::class, function (Server $c) {
 			$config = $c->getConfig();
-			$factoryClass = $config->getSystemValue('sharing.managerFactory', '\OC\Share20\ProviderFactory');
+			$factoryClass = $config->getSystemValue('sharing.managerFactory', Share20\ProviderFactory::class);
```
Comment (ChristophWurst) [inline lib/private/Server.php]: same, and also those above

**Review by ChristophWurst — DISMISSED** [review]

File: lib/private/Server.php (hunk 3 above)
Comment (MorrisJobke) [reply lib/private/Server.php]: There are multiple ManagerFactories. For example for the comments ;)

File: lib/private/Server.php (hunk 4 above)
Comment (MorrisJobke) [reply lib/private/Server.php]: Same here ;)

File: lib/private/Server.php (hunk 5 above)
Comment (MorrisJobke) [reply lib/private/Server.php]: Same here ;)

File: lib/private/Server.php (hunk 3 above)
Comment (ChristophWurst) [reply lib/private/Server.php]: ``use …\SystemTag\ManagerFactory as SystemTagManagerFactory`` 😉

Comment (ChristophWurst) [reply lib/private/Server.php]: or alternatively use the fully qualified class name, starting with `\OC\…`

Comment (nickvergessen) [reply lib/private/Server.php]: At least the import of `OC` should be killed...

**Comment by MorrisJobke** [comment]
I also reported and issue I found with the inspection: https://github.com/kalessil/phpinspectionsea/issues/770

File: lib/private/Server.php (hunk 3 above)
Comment (MorrisJobke) [reply lib/private/Server.php]: > At least the import of `OC` should be killed...

That is the namespace 😉 

Let me do the aliased import then.

**Comment by MorrisJobke** [comment]
Okay - it wasn't that easy. The problem was that with our event emitter we use classnames, but often we use a leading slash in the string. So the `::class` version would be without this leading slash which leads to hooks being not fired correctly and thus to failures in our unit tests. I fixed that now 🙈

**Comment by MorrisJobke** [comment]
Forgot to commit some local changes and wondered, why the tests pass locally but not on drone 🙈 Should work now.

**Review by rullzer — APPROVED** [review]

**Comment by juliushaertl** [comment]
Wow, those failing tests look pretty weird.

```
There were 117 failures:
1) Test\Share\HelperTest::testSplitUserRemote with data set #108 ('username@localhost', 'userlame', 'localhost')
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'userlame'
+'username'
```

**Comment by MorrisJobke** [comment]
All tests run totally fine locally. Also I could not find any occurrence of `userlame` in the code :/

Let me just rebase and check the result then.

**Comment by MorrisJobke** [comment]
> Let me just rebase and check the result then.

Now the tests pass.

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**